core/src/stages/generate_random_pose.cpp

```diff
@@ -0,0 +1,158 @@
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Authors: Henning Kayser */
+
+#include <moveit/task_constructor/stages/generate_random_pose.h>
+#include <moveit/task_constructor/storage.h>
+#include <moveit/task_constructor/marker_tools.h>
+#include <moveit/planning_scene/planning_scene.h>
+#include <rviz_marker_tools/marker_creation.h>
+
+#include <Eigen/Geometry>
+#include <tf2_eigen/tf2_eigen.hpp>
+
+#include <chrono>
+
+namespace {
+// TODO(henningkayser): support user-defined random number engines
+std::random_device RD;
+std::mt19937 GEN(RD());
+}  // namespace
+
+namespace moveit::task_constructor::stages {
+
+GenerateRandomPose::GenerateRandomPose(const std::string& name) : GeneratePose(name) {
+	auto& p = properties();
+	p.declare<geometry_msgs::msg::PoseStamped>("pose", "target pose to pass on in spawned states");
+	p.declare<size_t>("max_solutions", 20,
+	                  "maximum number of spawned solutions in case randomized sampling is enabled");
+	p.property("timeout").setDefaultValue(1.0 /* seconds */);
+}
+
+template <>
+GenerateRandomPose::PoseDimensionSampler
+GenerateRandomPose::getPoseDimensionSampler<std::normal_distribution>(double stddev) {
+	return [stddev, &gen = GEN](double mean) {
+		static std::normal_distribution<double> dist(mean, stddev);
+		return dist(gen);
+	};
+}
+
+template <>
+GenerateRandomPose::PoseDimensionSampler
+GenerateRandomPose::getPoseDimensionSampler<std::uniform_real_distribution>(double range) {
+	return [range, &gen = GEN](double mean) {
+		static std::uniform_real_distribution<double> dist(mean - 0.5 * range, mean + 0.5 * range);
+		return dist(gen);
+	};
+}
+
+bool GenerateRandomPose::canCompute() const {
+	return GeneratePose::canCompute() && !pose_dimension_samplers_.empty();
+}
+
+void GenerateRandomPose::compute() {
+	if (upstream_solutions_.empty())
+		return;
+
+	planning_scene::PlanningScenePtr scene = upstream_solutions_.pop()->end()->scene()->diff();
+	geometry_msgs::msg::PoseStamped target_pose = properties().get<geometry_msgs::msg::PoseStamped>("pose");
+	if (target_pose.header.frame_id.empty())
+		target_pose.header.frame_id = scene->getPlanningFrame();
```

Comment:
What do you think about this printing a warning that the user did not set the target frame_id and we are defaulting to the planning frame? Seems weird that we took in a PoseStamped with missing information.